### PR TITLE
[Serialization] Don't fail the precise tag check if only the last digit doesn't match

### PR DESCRIPTION
**Pull request by xymus**
Weaken the precise tag check at loading swiftmodules to accept binary modules build by a compiler with a tag where only the last digit is different. This relies on the assumption that the other digits in the version should ensure compiler and stdlib compatibility. If the last digit doesn't match, raise a remark.

The new `getCurrentCompilerSerializationTag` can be used to add vendor-specific information in the tag.

The precise tag check can be disabled with the env var `SWIFT_IGNORE_SWIFTMODULE_REVISION`.

rdar://105158258

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Review by artemcm — APPROVED** [review]

**Comment by xymus** [comment]
@swift-ci Please smoke test Linux

**Review by nkcsgexi — APPROVED** [review]
Thank you for incorporating this use case🙏

File: lib/Serialization/ModuleFileSharedCore.cpp [hunk 1]
```diff
@@ -370,12 +370,20 @@ static ValidationInfo validateControlBlock(
       if (forcedDebugRevision ||
           (requiresRevisionMatch && version::isCurrentCompilerTagged())) {
         StringRef compilerRevision = forcedDebugRevision ?
-          forcedDebugRevision : version::getCurrentCompilerTag();
+          forcedDebugRevision : version::getCurrentCompilerSerializationTag();
         if (moduleRevision != compilerRevision) {
-          result.status = Status::RevisionIncompatible;
+          // The module versions are mismatching, record it and diagnose later.
+          result.problematicRevision = moduleRevision;
```
Comment (bnbarham) [inline lib/Serialization/ModuleFileSharedCore.cpp:376]: Could go with `mismatchedRevision`, but `problematic` is okay too

File: include/swift/Serialization/Validation.h [hunk 2]
```diff
@@ -91,6 +91,7 @@ struct ValidationInfo {
   version::Version compatibilityVersion = {};
   llvm::VersionTuple userModuleVersion;
   StringRef sdkName = {};
+  StringRef problematicRevision = {};
```
Comment (bnbarham) [inline include/swift/Serialization/Validation.h:94]: All the `{}` aren't actually needed here - they'd already all be default initialized (like `VersionTuple`).

**Review by bnbarham — APPROVED** [review]

**Comment by xymus** [comment]
@swift-ci Please smoke test macos

File: lib/Serialization/ModuleFileSharedCore.cpp (hunk 1 above)
Comment (xymus) [reply lib/Serialization/ModuleFileSharedCore.cpp:376]: I was working in diagnostics at the same time where we do use the generic `problematic` but here `mismatched` would be more precise. I'll try to update it next time I work in this code.

**Pull request merged**